File: lib/features/chat/data/data/models/message_model.dart
import 'package:chat_application_task/features/chat/domain/entities/message_entity.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
part 'message_model.freezed.dart';
part 'message_model.g.dart';

@freezed
abstract class MessageModel with _$MessageModel {
  const factory MessageModel({
    required String id,
    required String chatId,
    required String senderId,
    required String receiverId,
    required String content,
    required DateTime timestamp,
    required String status,
  }) = _MessageModel;

  factory MessageModel.fromJson(Map<String, dynamic> json) =>
      _$MessageModelFromJson(json);
}

extension MessageModelX on MessageModel {
 MessageEntity toMessageEntity() {
    return MessageEntity(
      id: id,
      senderId: senderId,
      receiverId: receiverId,
      chatId: chatId,
      content: content,
      timestamp: timestamp,
      status: MessageStatus.values.firstWhere((e) => e.name == status),
    );
  }
}

File: android/app/src/main/kotlin/com/example/chat_application_task/models/MessageModel.kt
package com.example.chat_application_task.models

data class MessageModel(
    val messageId: String,
    val chatId: String,
    val senderId: String,
    val receiverId: String,
    val content: String,
    val status: String,
    val timestamp: String,
){
    companion object {
        fun fromMap(data: Map<String, Any>): MessageModel {
            return MessageModel(
                messageId = (data["id"] ?: data["messageId"]) as String,
                chatId = data["chatId"] as String,
                senderId = data["senderId"] as String,
                receiverId = data["receiverId"] as String,
                content = data["content"] as String,
                status = (data["status"] as? String) ?: "sent",
                timestamp = (data["timestamp"] as String),
            )
        }

        fun isAnyDataMissing(data: Map<String, Any>): Boolean {
            val requiredKeys = listOf("chatId", "senderId", "receiverId", "content", "timestamp")
            return requiredKeys.any { key -> !data.containsKey(key) }
        }
    }
}